hoomd/box.py

```diff
@@ -0,0 +1,402 @@
+    saved :py:class:`State` variable (``state.box = new_box``) to set the
+    simulation box.
+
+    Access attributes directly::
+
+        b = hoomd.box.Box(L=20)
+        b.xy = 1.0
+        b.yz = 0.5
+        b.Lz = 40
+
+    .. rubric:: Two dimensional systems
+
+    2D simulations in HOOMD use boxes with ``Lz == 0``. 2D boxes ignore ``xz``
+    and ``yz``. If a new `Box` is assigned to a system with differnt
+    dimensionality, a warning will be shown.
+
+    In 2D boxes, *volume* is in units of area.
+
+    .. rubric:: Factory Methods
+
+    `Box` has factory methods to enable easier creation of boxes:
+    `cube` and `from_matrix`. See the method documentation for usage.
+
+    Examples:
+
+    * Cubic box with given length: ``Box.cube(L=1)``
+    * Square box with given length: ``Box.cube(L=1, dimensions=2)``
+    * From an upper triangular matrix: ``Box.from_matrix(matrix)``
+    * Full spec: ``Box(Lx=1, Ly=2, Lz=3, xy=1., xz=2., yz=3.)``
+    """
+
+    # Constructors
+    def __init__(self, Lx, Ly, Lz=0, xy=0, xz=0, yz=0):
+        self._cpp_obj = _hoomd.BoxDim(Lx, Ly, Lz)
+        self._cpp_obj.setTiltFactors(xy, xz, yz)
+
+    @classmethod
+    def cube(cls, L):
+        """Create a cube with side lengths ``L``.
+
+        Args:
+            L (float): The box side length (distance units).
+
+        Returns:
+            box (hoomd.Box): The created box.
+        """
+        return cls(L, L, L, 0, 0, 0)
+
+    @classmethod
+    def square(cls, L):
+        """Create a square with side lengths ``L``.
+
+        Args:
+            L (float): The box side length (distance units).
+
+        Returns:
+            box (hoomd.Box): The created 2D box.
+        """
+        return cls(L, L, 0, 0, 0, 0)
+
+    @classmethod
+    def from_matrix(cls, box_matrix):
+        """Create a box from an upper triangular matrix.
+
+        Args:
+            box_matrix ((3, 3) `numpy.ndarray` of `float`): An upper
+                triangular matrix representing a box. The values for ``Lx``,
+                ``Ly``, ``Lz``, ``xy``, ``xz``, and ``yz`` are related to the
+                matrix by the following expressions.
+
+                .. code-block:: python
+
+                    [[Lx, Ly * xy, Lz * xz],
+                    [0,  Ly,      Lz * yz],
+                    [0,  0,       Lz]]
+
+
+        Returns:
+            box (hoomd.Box): The created box.
+        """
+        b = cls(0, 0)
+        b.matrix = box_matrix
+        return b
+
+    @classmethod
+    def _from_cpp(self, cpp_obj):
+        b = Box()
+        b._cpp_obj = cpp_obj
+        return b
+
+    # Dimension based properties
+    @property
+    def dimensions(self):
+        """int: The dimensionality of the box.
+
+        If ``Lz == 0``, the box is treated as 2D, otherwise it is 3D. This
+        property is not settable.
+        """
+        return 2 if self.is2D else 3
+
+    @property
+    def is2D(self):
```

Comment:
What's the purpose of having both this and dimensions in the public API?

Reply:
`freud` does this. It is an easy conversion from one to the other, but sometimes one is more handy than the other.

Reply:
How is one more handy than the other if the one just calls the other? Especially if `is2D` is only called in the dimension property and nowhere else in the code base?

Reply:
I more meant for users. `box.is2D` is nice for conditionals while `box.dimensions` is useful for certain math calculations.

Reply:
I can attest to this. Sometimes you want to do something like `if box.is2d: do stuff`, whereas at other times you might want to do e.g. `num_particles=num_replicas**box.dimensions`.